File: scripts/strict_qa.py
#!/usr/bin/env python3
"""Strict QA validation for the world knowledge runtime."""

from __future__ import annotations

import argparse
import json
from typing import Any

from knowledge_store_common import (
    VALID_CURRICULUM_ROLES,
    VALID_DOMAINS,
    VALID_EDGE_TYPES,
    VALID_LEARNING_MODES,
    VALID_NODE_KINDS,
    VALID_SCHOOL_STAGES,
    connect_db,
    ensure_pg_schema,
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Strict QA for world knowledge graph.")
    parser.add_argument("--dataset-id", required=True)
    parser.add_argument("--db")
    return parser.parse_args()


class StrictQA:
    def __init__(self, connection, dataset_id: str):
        self.connection = connection
        self.dataset_id = dataset_id
        self.errors: list[dict[str, Any]] = []
        self.warnings: list[dict[str, Any]] = []
        self._evidence_exists_cache: dict[str, bool] = {}

    def error(self, category: str, item_id: str, message: str) -> None:
        self.errors.append({"category": category, "id": item_id, "message": message})

    def warn(self, category: str, item_id: str, message: str) -> None:
        self.warnings.append({"category": category, "id": item_id, "message": message})

    def evidence_exists(self, evidence_id: str) -> bool:
        if evidence_id not in self._evidence_exists_cache:
            with self.connection.cursor() as cur:
                cur.execute(
                    "SELECT 1 FROM world_evidence WHERE dataset_id = %s AND id = %s",
                    (self.dataset_id, evidence_id),
                )
                self._evidence_exists_cache[evidence_id] = cur.fetchone() is not None
        return self._evidence_exists_cache[evidence_id]

    def validate_source_refs(self, category: str, item_id: str, source_refs: Any, *, required: bool = True) -> None:
        refs = source_refs if isinstance(source_refs, list) else []
        if required and not refs:
            self.error(category, item_id, "Missing evidence source references")
            return
        for evidence_id in refs:
            if not isinstance(evidence_id, str) or not evidence_id.strip():
                self.error(category, item_id, "Invalid empty evidence reference")
                continue
            if not self.evidence_exists(evidence_id):
                self.error(category, item_id, f"Missing evidence {evidence_id}")

    def validate_nodes(self) -> None:
        with self.connection.cursor() as cur:
            cur.execute("SELECT * FROM world_nodes WHERE dataset_id = %s", (self.dataset_id,))
            rows = cur.fetchall()
        for row in rows:
            if row["kind"] not in VALID_NODE_KINDS:
                self.error("node", row["id"], f"Invalid kind: {row['kind']}")
            if not row["name"] or not row["definition"]:
                self.error("node", row["id"], "Missing name or definition")
            if not isinstance(row["domains_json"], list) or not row["domains_json"]:
                self.error("node", row["id"], "domains_json must be a non-empty array")
            else:
                invalid = [item for item in row["domains_json"] if item not in VALID_DOMAINS]
                if invalid:
                    self.error("node", row["id"], f"Invalid domains: {invalid}")
            if not isinstance(row["learning_mode_json"], list) or not row["learning_mode_json"]:
                self.error("node", row["id"], "learning_mode_json must be a non-empty array")
            else:
                invalid = [item for item in row["learning_mode_json"] if item not in VALID_LEARNING_MODES]
                if invalid:
                    self.error("node", row["id"], f"Invalid learning modes: {invalid}")
            with self.connection.cursor() as cur:
                cur.execute("SELECT 1 FROM world_node_cards WHERE dataset_id = %s AND node_id = %s", (self.dataset_id, row["id"]))
                if cur.fetchone() is None:
                    self.error("node_card", row["id"], "Missing node card")
                cur.execute("SELECT 1 FROM world_mentions WHERE dataset_id = %s AND target_id = %s", (self.dataset_id, row["id"]))
                if cur.fetchone() is None:
                    self.error("mention", row["id"], "Missing mention")
                cur.execute("SELECT 1 FROM world_domain_profiles WHERE dataset_id = %s AND node_id = %s", (self.dataset_id, row["id"]))
                if cur.fetchone() is None:
                    self.error("domain_profile", row["id"], "Missing domain profile")

    def validate_edges(self) -> None:
        with self.connection.cursor() as cur:
            cur.execute("SELECT * FROM world_edges WHERE dataset_id = %s", (self.dataset_id,))
            rows = cur.fetchall()
        for row in rows:
            if row["type"] not in VALID_EDGE_TYPES:
                self.error("edge", row["id"], f"Invalid edge type: {row['type']}")
            if row["directionality"] not in {"directed", "undirected"}:
                self.error("edge", row["id"], f"Invalid directionality: {row['directionality']}")
            self.validate_source_refs("edge", row["id"], row["source_refs_json"])
            with self.connection.cursor() as cur:
                cur.execute("SELECT 1 FROM world_nodes WHERE dataset_id = %s AND id = %s", (self.dataset_id, row["from_id"]))
                if cur.fetchone() is None:
                    self.error("edge", row["id"], "Missing source node")
                cur.execute("SELECT 1 FROM world_nodes WHERE dataset_id = %s AND id = %s", (self.dataset_id, row["to_id"]))
                if cur.fetchone() is None:
                    self.error("edge", row["id"], "Missing target node")

    def validate_domain_profiles(self) -> None:
        with self.connection.cursor() as cur:
            cur.execute("SELECT * FROM world_domain_profiles WHERE dataset_id = %s", (self.dataset_id,))
            rows = cur.fetchall()
        for row in rows:
            if row["domain"] not in VALID_DOMAINS:
                self.error("domain_profile", row["id"], f"Invalid domain: {row['domain']}")
            invalid_stages = [item for item in (row["school_stages_json"] or []) if item not in VALID_SCHOOL_STAGES]
            if invalid_stages:
                self.error("domain_profile", row["id"], f"Invalid school stages: {invalid_stages}")
            invalid_roles = [item for item in (row["curriculum_roles_json"] or []) if item not in VALID_CURRICULUM_ROLES]
            if invalid_roles:
                self.error("domain_profile", row["id"], f"Invalid curriculum roles: {invalid_roles}")
            self.validate_source_refs("domain_profile", row["id"], row["source_refs_json"])

    def validate_mentions_and_evidence(self) -> None:
        with self.connection.cursor() as cur:
            cur.execute("SELECT * FROM world_mentions WHERE dataset_id = %s", (self.dataset_id,))
            mentions = cur.fetchall()
        for row in mentions:
            self.validate_source_refs("mention", row["id"], row["source_refs_json"])

    def validate_node_cards(self) -> None:
        with self.connection.cursor() as cur:
            cur.execute("SELECT * FROM world_node_cards WHERE dataset_id = %s", (self.dataset_id,))
            cards = cur.fetchall()
        required_sections = {"definition", "essence", "key_points", "example", "application", "misconception"}
        for row in cards:
            if not row["summary"]:
                self.error("node_card", row["node_id"], "Missing summary")
            self.validate_source_refs("node_card", row["node_id"], row["source_refs_json"])
            sections = row["sections_json"] if isinstance(row["sections_json"], list) else []
            section_types = {section.get("section_type") for section in sections if isinstance(section, dict)}
            missing = sorted(required_sections - section_types)
            if missing:
                self.error("node_card", row["node_id"], f"Missing required sections: {missing}")
            for section in sections:
                if not isinstance(section, dict):
                    continue
                section_id = str(section.get("id") or section.get("section_type") or "section")
                self.validate_source_refs(
                    "node_card_section",
                    f"{row['node_id']}:{section_id}",
                    section.get("source_refs"),
                )

    def run(self) -> dict[str, Any]:
        self.validate_nodes()
        self.validate_edges()
        self.validate_domain_profiles()
        self.validate_mentions_and_evidence()
        self.validate_node_cards()
        return {"errors": self.errors, "warnings": self.warnings}


def main() -> int:
    args = parse_args()
    connection = connect_db(args.db)
    ensure_pg_schema(connection)
    qa = StrictQA(connection, args.dataset_id)
    result = qa.run()
    status = "success" if not result["errors"] else "blocked"
    print(json.dumps({"status": status, **result}, ensure_ascii=False))
    return 0 if status == "success" else 2


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/pipeline/src/qa/strict-qa.ts
import {
  VALID_CURRICULUM_ROLES,
  VALID_DOMAINS,
  VALID_EDGE_TYPES,
  VALID_LEARNING_MODES,
  VALID_NODE_KINDS,
  VALID_SCHOOL_STAGES,
} from "../shared/knowledge.js";

export type StrictQaIssue = {
  category: string;
  id: string;
  message: string;
};

export type StrictQaRows = {
  nodes: Array<{
    id: string;
    kind: string;
    name?: string | null;
    definition?: string | null;
    domains_json?: unknown;
    learning_mode_json?: unknown;
  }>;
  edges: Array<{
    id: string;
    type: string;
    directionality: string;
    from_id: string;
    to_id: string;
    source_refs_json?: unknown;
  }>;
  domain_profiles: Array<{
    id: string;
    node_id: string;
    domain: string;
    school_stages_json?: unknown;
    curriculum_roles_json?: unknown;
    source_refs_json?: unknown;
  }>;
  mentions: Array<{
    id: string;
    target_id: string;
    source_refs_json?: unknown;
  }>;
  evidence: Array<{
    id: string;
  }>;
  node_cards: Array<{
    node_id: string;
    summary?: string | null;
    source_refs_json?: unknown;
    sections_json?: unknown;
  }>;
};

export type StrictQaResult = {
  status: "success" | "blocked";
  errors: StrictQaIssue[];
  warnings: StrictQaIssue[];
};

const REQUIRED_CARD_SECTIONS = new Set(["definition", "essence", "key_points", "example", "application", "misconception"]);

export function runStrictQa(rows: StrictQaRows): StrictQaResult {
  const qa = new StrictQaRunner(rows);
  const result = qa.run();
  return {
    status: result.errors.length === 0 ? "success" : "blocked",
    ...result,
  };
}

class StrictQaRunner {
  private readonly errors: StrictQaIssue[] = [];
  private readonly warnings: StrictQaIssue[] = [];
  private readonly nodeIds: Set<string>;
  private readonly evidenceIds: Set<string>;
  private readonly cardNodeIds: Set<string>;
  private readonly mentionTargetIds: Set<string>;
  private readonly profileNodeIds: Set<string>;

  constructor(private readonly rows: StrictQaRows) {
    this.nodeIds = new Set(rows.nodes.map((row) => row.id));
    this.evidenceIds = new Set(rows.evidence.map((row) => row.id));
    this.cardNodeIds = new Set(rows.node_cards.map((row) => row.node_id));
    this.mentionTargetIds = new Set(rows.mentions.map((row) => row.target_id));
    this.profileNodeIds = new Set(rows.domain_profiles.map((row) => row.node_id));
  }

  run(): Omit<StrictQaResult, "status"> {
    this.validateNodes();
    this.validateEdges();
    this.validateDomainProfiles();
    this.validateMentionsAndEvidence();
    this.validateNodeCards();
    return { errors: this.errors, warnings: this.warnings };
  }

  private error(category: string, id: string, message: string): void {
    this.errors.push({ category, id, message });
  }

  private validateSourceRefs(category: string, itemId: string, sourceRefs: unknown, options: { required?: boolean } = {}): void {
    const required = options.required ?? true;
    const refs = Array.isArray(sourceRefs) ? sourceRefs : [];
    if (required && refs.length === 0) {
      this.error(category, itemId, "Missing evidence source references");
      return;
    }
    for (const evidenceId of refs) {
      if (typeof evidenceId !== "string" || !evidenceId.trim()) {
        this.error(category, itemId, "Invalid empty evidence reference");
        continue;
      }
      if (!this.evidenceIds.has(evidenceId)) {
        this.error(category, itemId, `Missing evidence ${evidenceId}`);
      }
    }
  }

  private validateNodes(): void {
    for (const row of this.rows.nodes) {
      if (!VALID_NODE_KINDS.has(row.kind)) {
        this.error("node", row.id, `Invalid kind: ${row.kind}`);
      }
      if (!row.name || !row.definition) {
        this.error("node", row.id, "Missing name or definition");
      }
      const domains = Array.isArray(row.domains_json) ? row.domains_json : null;
      if (!domains || domains.length === 0) {
        this.error("node", row.id, "domains_json must be a non-empty array");
      } else {
        const invalid = domains.filter((item) => !VALID_DOMAINS.has(String(item)));
        if (invalid.length > 0) this.error("node", row.id, `Invalid domains: ${formatPythonStringList(invalid)}`);
      }
      const learningModes = Array.isArray(row.learning_mode_json) ? row.learning_mode_json : null;
      if (!learningModes || learningModes.length === 0) {
        this.error("node", row.id, "learning_mode_json must be a non-empty array");
      } else {
        const invalid = learningModes.filter((item) => !VALID_LEARNING_MODES.has(String(item)));
        if (invalid.length > 0) this.error("node", row.id, `Invalid learning modes: ${formatPythonStringList(invalid)}`);
      }
      if (!this.cardNodeIds.has(row.id)) this.error("node_card", row.id, "Missing node card");
      if (!this.mentionTargetIds.has(row.id)) this.error("mention", row.id, "Missing mention");
      if (!this.profileNodeIds.has(row.id)) this.error("domain_profile", row.id, "Missing domain profile");
    }
  }

  private validateEdges(): void {
    for (const row of this.rows.edges) {
      if (!VALID_EDGE_TYPES.has(row.type)) this.error("edge", row.id, `Invalid edge type: ${row.type}`);
      if (row.directionality !== "directed" && row.directionality !== "undirected") this.error("edge", row.id, `Invalid directionality: ${row.directionality}`);
      this.validateSourceRefs("edge", row.id, row.source_refs_json);
      if (!this.nodeIds.has(row.from_id)) this.error("edge", row.id, "Missing source node");
      if (!this.nodeIds.has(row.to_id)) this.error("edge", row.id, "Missing target node");
    }
  }

  private validateDomainProfiles(): void {
    for (const row of this.rows.domain_profiles) {
      if (!VALID_DOMAINS.has(row.domain)) this.error("domain_profile", row.id, `Invalid domain: ${row.domain}`);
      const invalidStages = (Array.isArray(row.school_stages_json) ? row.school_stages_json : []).filter((item) => !VALID_SCHOOL_STAGES.has(String(item)));
      if (invalidStages.length > 0) this.error("domain_profile", row.id, `Invalid school stages: ${formatPythonStringList(invalidStages)}`);
      const invalidRoles = (Array.isArray(row.curriculum_roles_json) ? row.curriculum_roles_json : []).filter((item) => !VALID_CURRICULUM_ROLES.has(String(item)));
      if (invalidRoles.length > 0) this.error("domain_profile", row.id, `Invalid curriculum roles: ${formatPythonStringList(invalidRoles)}`);
      this.validateSourceRefs("domain_profile", row.id, row.source_refs_json);
    }
  }

  private validateMentionsAndEvidence(): void {
    for (const row of this.rows.mentions) {
      this.validateSourceRefs("mention", row.id, row.source_refs_json);
    }
  }

  private validateNodeCards(): void {
    for (const row of this.rows.node_cards) {
      if (!row.summary) this.error("node_card", row.node_id, "Missing summary");
      this.validateSourceRefs("node_card", row.node_id, row.source_refs_json);
      const sections = Array.isArray(row.sections_json) ? row.sections_json : [];
      const sectionTypes = new Set(sections.filter(isRecord).map((section) => section.section_type));
      const missing = [...REQUIRED_CARD_SECTIONS].filter((section) => !sectionTypes.has(section)).sort();
      if (missing.length > 0) this.error("node_card", row.node_id, `Missing required sections: ${formatPythonStringList(missing)}`);
      for (const section of sections) {
        if (!isRecord(section)) continue;
        const sectionId = String(section.id || section.section_type || "section");
        this.validateSourceRefs("node_card_section", `${row.node_id}:${sectionId}`, section.source_refs);
      }
    }
  }
}

function formatPythonStringList(values: unknown[]): string {
  return `[${values.map((value) => (typeof value === "string" ? `'${value}'` : String(value))).join(", ")}]`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}
